rust/src/response_broker.rs

```diff
@@ -0,0 +1,89 @@
+use futures_util::FutureExt;
+use std::collections::HashMap;
+use std::fmt::Debug;
+use std::future::Future;
+use std::sync::Mutex;
+use tokio::sync::oneshot;
+use tokio::sync::oneshot::error::RecvError;
+use tokio::sync::oneshot::Sender;
+use tracing::{error, debug};
+
+#[derive(Debug)]
+pub(crate) struct ResponseBroker<T: Clone + Debug, E: std::error::Error + Clone + From<RecvError>> {
+    responses: Mutex<HashMap<String, Sender<Result<T, E>>>>,
+}
+
+impl<T: Clone + Debug, E: From<RecvError> + Clone + std::error::Error> ResponseBroker<T, E> {
+    pub fn new() -> Self {
+        Self {
+            responses: Mutex::new(HashMap::new()),
+        }
+    }
+
+    pub fn get_response(&self, request_id: String) -> impl Future<Output = Result<T, E>> {
```

Comment:
NIT: Maybe `get_receiver` or `get_response_receiver` are better names? We are not getting a response back here.

Reply:
Yeah we also discussed the naming here. This choice was made to "hide" the underlying implementation from the caller. The fact that this method actually uses a channel to receive the response is an implementation detail, not a feature of the function. The function's contract just returns a future that will (eventually) yield a response.